Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,886 @@
+      Contract.Requires(tok != null);
+      Contract.Requires(parameters != null);
+      Contract.Requires(arguments != null);
+      Contract.Requires(parameters.Count == arguments.Count);
+
+      if (parameters.All(tp => tp.Variance == TypeParameter.TPVariance.Non)) {
+        // special case this common situation
+        return arguments;
+      }
+      var newArgs = new List<PreType>();
+      for (var i = 0; i < parameters.Count; i++) {
+        var tp = parameters[i];
+        if (tp.Variance == TypeParameter.TPVariance.Non) {
+          newArgs.Add(arguments[i]);
+        } else {
+          var co = tp.Variance == TypeParameter.TPVariance.Co ? "co" : "contra";
+          var proxy = CreatePreTypeProxy($"type used in {co}variance constraint");
+          newArgs.Add(proxy);
+          if ((tp.Variance == TypeParameter.TPVariance.Co) == proxiesAreSupertypes) {
+            AddSubtypeConstraint(proxy, arguments[i], tok, "bad"); // TODO: improve error message
+          } else {
+            AddSubtypeConstraint(arguments[i], proxy, tok, "bad"); // TODO: improve error message
+          }
+        }
+      }
+      return newArgs;
+    }
+
+    /// <summary>
+    /// For every non-variant parameters[i], constrain superArguments[i] == subArguments[i].
+    /// For every co-variant parameters[i], constrain superArguments[i] :> subArguments[i].
+    /// For every contra-variant parameters[i], constrain subArguments[i] :> superArguments[i].
+    /// </summary>
+    void ConstrainTypeArguments(List<TypeParameter> parameters, List<PreType> superArguments, List<PreType> subArguments, IToken tok) {
+      Contract.Requires(parameters != null);
+      Contract.Requires(superArguments != null);
+      Contract.Requires(subArguments != null);
+      Contract.Requires(parameters.Count == superArguments.Count && superArguments.Count == subArguments.Count);
+      Contract.Requires(tok != null);
+
+      for (var i = 0; i < parameters.Count; i++) {
+        var tp = parameters[i];
+        var arg0 = superArguments[i];
+        var arg1 = subArguments[i];
+        if (tp.Variance == TypeParameter.TPVariance.Non) {
+          AddEqualityConstraint(arg0, arg1, tok, "non-variance would require {0} == {1}");
+        } else if (tp.Variance == TypeParameter.TPVariance.Co) {
+          AddSubtypeConstraint(arg0, arg1, tok, "covariance would require {0} :> {1}");
+        } else {
+          AddSubtypeConstraint(arg1, arg0, tok, "contravariance would require {0} :> {1}");
+        }
+      }
+    }
+
+    bool DecideHeadsFromBounds(bool fromSubBounds) {
+      // For each proxy, compute the join/meet of its sub/super-bound heads
+      Dictionary<PreTypeProxy, TopLevelDecl> candidateHeads = new();
+      Dictionary<PreTypeProxy, SubtypeConstraint> constraintOrigins = new();
+      foreach (var constraint in unnormalizedSubtypeConstraints) {
```

Comment:
Subtype and equality constraints are processed before "guarded constraints", so there can't be just a single `constraints` queue. It could be a priority queue.

There is one other consideration, which is that, sometimes, _all_ of the subset constraints are searched when trying to find a bound for a proxy pre-type. This makes subset constraints special.

It's entirely possible that there's a better data structure (like the separate variables for upper/lower bounds) that would be more efficient. For now, the simple lists are used.

Reply:
For the sake of time we can leave it as is, but my intuition is that it would both be easier to understand and more performant if all of the iterations over the subset constraints would be replaced with separate state.